File: dotnet/ch4/documents/ParseWordDocument.cs
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;
using System.ComponentModel;
using Microsoft.SemanticKernel;
namespace Plugins.ProposalChecker;
public class ParseWordDocument
{

    [KernelFunction, Description("Extracts the text under the Team heading in the Word document")]
    public static string ExtractTeam(string folderPath)
    {
        if (folderPath.Contains("Error"))
        {
            return folderPath;
        }        
        string text = ExtractTextUnderHeading(folderPath, "Team");
        return $"FolderPath: {folderPath}\n"  + text;

    }

    [KernelFunction, Description("Extracts the text under the Experience heading in the Word document")]
    public static string ExtractExperience(string folderPath)
    {
        if (folderPath.Contains("Error"))
        {
            return folderPath;
        }            
        string text = ExtractTextUnderHeading(folderPath, "Experience");
        return $"FolderPath: {folderPath}\n"  + text;
    }

    [KernelFunction, Description("Extracts the text under the Implementation heading in the Word document")]
    public static string ExtractImplementation(string folderPath)
    {
        if (folderPath.Contains("Error"))
        {
            return folderPath;
        }            
        string text = ExtractTextUnderHeading(folderPath, "Implementation");
        return $"FolderPath: {folderPath}\n"  + text;
    }

    private static string ExtractTextUnderHeading(string folderPath, string heading)
    {
        if (folderPath.StartsWith("FolderPath:"))
        {
            folderPath = folderPath.Substring("FolderPath: ".Length);
            folderPath = folderPath.Trim();
        }
        string filePath = Directory.GetFiles(folderPath).First(f => Path.GetExtension(f).ToLower() == ".docx");

        using (WordprocessingDocument doc = WordprocessingDocument.Open(filePath, false))
        {
            // 
            var body = doc.MainDocumentPart?.Document.Body;
            var paras = body?.Elements<Paragraph>();

            bool isExtracting = false;
            string extractedText = "";

            if (paras == null)
            {
                return $"Error: Missing section {heading}";
            }

            foreach (var para in paras)
            {
                if (para.ParagraphProperties != null &&
                    para.ParagraphProperties.ParagraphStyleId != null &&
                    para.ParagraphProperties.ParagraphStyleId.Val != null &&
                    para.ParagraphProperties.ParagraphStyleId.Val.Value == "Heading1" &&
                    para.InnerText.Trim().Equals(heading, StringComparison.OrdinalIgnoreCase))
                {
                    isExtracting = true;
                    continue;
                }

                if (isExtracting)
                {
                    if (para.ParagraphProperties != null &&
                    para.ParagraphProperties.ParagraphStyleId != null &&
                    para.ParagraphProperties.ParagraphStyleId.Val != null &&
                        para.ParagraphProperties.ParagraphStyleId.Val.Value == "Heading1")
                    {
                        break;
                    }

                    extractedText += para.InnerText + "\n";
                }
            }

            if (extractedText == "")
            {
                return $"Error: Missing section {heading}";
            }

            return extractedText.Trim();
        }
    }
}

File: python/ch4/ParseWordDocument.py
from docx import Document
from typing_extensions import Annotated
from semantic_kernel.functions.kernel_function_decorator import kernel_function
import os

class ParseWordDocument:

    @kernel_function(
        description="Extracts information about the team",
        name="ExtractTeam",
    )
    def ExtractTeam(self,
                    input: Annotated[str, "The path of the file to extract the text from"]) ->  \
                    Annotated[str, "The extracted text about the team description text"]:                    
        if input.startswith("Error"):
            return input
        doc_path = self.get_first_docx_file(input)
        text = self.ExtractTextUnderHeading(doc_path, 'Team')
        return f"FolderPath: {input}\n{text}"

    @kernel_function(
        description="Extracts information about the team's experience on their previous project",
        name="ExtractExperience",
    )
    def ExtractExperience(self,
                              input: Annotated[str, "The path of the file to extract the text from"]) ->  \
                              Annotated[str, "The extracted experience text"]:                          
        if input.startswith("Error"):
            return input
        doc_path = self.get_first_docx_file(input)
        text =  self.ExtractTextUnderHeading(doc_path, 'Experience')
        return f"FolderPath: {input}\n{text}"
    
    @kernel_function(
        description="Extracts information about the implementation plan being proposed",
        name="ExtractImplementation",
    )
    def ExtractImplementation(self, 
                              input: Annotated[str, "The path of the file to extract the text from"]) ->  \
                              Annotated[str, "The extracted implementation text"]:
        if input.startswith("Error"):
            return input
        doc_path = self.get_first_docx_file(input)
        text = self.ExtractTextUnderHeading(doc_path, 'Implementation')
        return f"FolderPath: {input}\n{text}"

    def get_first_docx_file(self, input: str) -> str:
        if input.startswith("FolderPath"):
            input = input.split(":")[1].strip()
        if input.startswith("Error"):
            return input
        for file in os.listdir(input):
            if file.endswith(".docx"):
                return os.path.join(input, file)

    def ExtractTextUnderHeading(self, doc_path: str, target_heading: str) -> str:
        doc = Document(doc_path)
        extract = False
        extracted_text = ''

        for paragraph in doc.paragraphs:
            if paragraph.style.name == 'Heading 1':
                if extract:
                    break  # Stop if next heading is found
                extract = paragraph.text.strip().lower() == target_heading.lower()
            elif extract:
                extracted_text += paragraph.text + '\n'
        return extracted_text.strip()
